### PR TITLE
Client: Initial Client Spider tab

### PR DESCRIPTION
**Pull request by psiinon**
## Overview

Add the Client Spider tab. No help as this is still a "hidden" feature.
Currently only the nodes added are shown. We should report the full set of URLs found but that will require other changes.

## Checklist
- [ ] Update help
- [ ] Update changelog
- [x] Run `./gradlew spotlessApply` for code formatting
- [ ] Write tests
- [x] Check code coverage
- [x] Sign-off commits
- [x] Squash commits
- [x] Use a descriptive title

For more details, please refer to the [developer rules and guidelines](https://www.zaproxy.org/docs/developer/dev-rules-and-guidelines/).

### CODE REVIEW

**Comment by psiinon** [comment]
Logo
**Checkmarx One – Scan Summary & Details** – [6e6cf0fe\-4c2c\-49c7\-a6df\-586dfdc5ab94](https://eu.ast.checkmarx.net/projects/0da811fa-6373-42b1-bbb6-d0026285b136/scans?id=6e6cf0fe-4c2c-49c7-a6df-586dfdc5ab94&branch=psiinon%2Fclient%2Fgui)
### New Issues
Severity|Issue|Source File / Package|Checkmarx Insight
---|---|---|---
LOW|[Log_Forging](https://eu.ast.checkmarx.net/results/0da811fa-6373-42b1-bbb6-d0026285b136/6e6cf0fe-4c2c-49c7-a6df-586dfdc5ab94/sast/description/117/9884244256835554923)|/addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpiderDialog\.java: [343](https://github.com/psiinon/zap-extensions/blob/client/gui//addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpiderDialog.java#L343)|[Attack Vector](https://eu.ast.checkmarx.net/results/6e6cf0fe-4c2c-49c7-a6df-586dfdc5ab94/0da811fa-6373-42b1-bbb6-d0026285b136/sast?result-id=k%2FyUzP6ue%2FLCBm222Cb%2FElvp8XY%3D&redirect=true)

File: addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java [hunk 1]
```diff
@@ -146,15 +157,17 @@ public void initModel(Model model) {
...
+        spiderScanController = new SpiderScanController(this);
+        passiveScanController = new ClientPassiveScanController();
     }
 
     @Override
     public void hook(ExtensionHook extensionHook) {
         super.hook(extensionHook);
 
-        scanController = new ClientPassiveScanController();
         this.api = new ClientIntegrationAPI(this);
```
Comment (kingthorin) [inline addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java:168]: `ClientIntegrationApi`

File: addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpiderPanel.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger LOGGER = LogManager.getLogger(ClientSpiderPanel.class);
+
+    private static final String ZERO_REQUESTS_LABEL_TEXT = "0";
+
+    private static final UrlTableModel EMPTY_TABLE_MODEL = new UrlTableModel();
+
+    /** The Constant defining the PANEL's NAME. */
```
Comment (kingthorin) [inline addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpiderPanel.java:61]: I think this comment can go 😄

File: addOns/client/src/main/java/org/zaproxy/addon/client/spider/UrlTableModel.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    /** Removes all the elements. Method is synchronized internally. */
+    public void removeAllElements() {
+        scanResults.clear();
+        fireTableDataChanged();
+    }
+
+    /**
+     * Adds a new spider scan result. Method is synchronized internally.
+     *
+     * @param uri the uri
```
Comment (kingthorin) [inline addOns/client/src/main/java/org/zaproxy/addon/client/spider/UrlTableModel.java:83]: "the uri of the result being added" ?

Comment (kingthorin) [reply addOns/client/src/main/java/org/zaproxy/addon/client/spider/UrlTableModel.java:83]: Same for all the methods below, comments saying the param is the param is not exactly useful

File: addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java (hunk 1 above)
Comment (psiinon) [reply addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java:168]: Thats existing code.

File: addOns/client/src/main/java/org/zaproxy/addon/client/spider/UrlTableModel.java (hunk 3 above)
Comment (psiinon) [reply addOns/client/src/main/java/org/zaproxy/addon/client/spider/UrlTableModel.java:83]: yeah, copy and paste code :P

File: addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java (hunk 1 above)
Comment (kingthorin) [reply addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java:168]: Shouldn't matter to anyone but us.

**Comment by psiinon** [comment]
Fixed conflict and some of the comments

File: addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java (hunk 1 above)
Comment (psiinon) [reply addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java:167]: Agreed, but thats not a change I introduced in this PR. Happy for anyone to rename it in another PR, but in general I dont think we should ask people to fix existing code in PRs unless its directly related to the changes being made.

**Comment by thc202** [comment]
Thank you!

**Review by kingthorin — APPROVED** [review]

**Pull request merged**